bfvmm/include/intrinsics/x86/intel/apic/x2apic.h

```diff
@@ -2060,36 +2134,183 @@ namespace x2apic
+            case idx3:
+                return msrs::ia32_x2apic_tmr3::get();
+            case idx4:
+                return msrs::ia32_x2apic_tmr4::get();
+            case idx5:
+                return msrs::ia32_x2apic_tmr5::get();
+            case idx6:
+                return msrs::ia32_x2apic_tmr6::get();
+            case idx7:
+                return msrs::ia32_x2apic_tmr7::get();
+
+            default:
+                return 0;
+        }
+    }
+
+    value_type read_irr(index idx) noexcept override
+    {
+        switch (idx) {
+            case idx0:
+                return msrs::ia32_x2apic_irr0::get();
+            case idx1:
+                return msrs::ia32_x2apic_irr1::get();
+            case idx2:
+                return msrs::ia32_x2apic_irr2::get();
+            case idx3:
+                return msrs::ia32_x2apic_irr3::get();
+            case idx4:
+                return msrs::ia32_x2apic_irr4::get();
+            case idx5:
+                return msrs::ia32_x2apic_irr5::get();
+            case idx6:
+                return msrs::ia32_x2apic_irr6::get();
+            case idx7:
+                return msrs::ia32_x2apic_irr7::get();
+
+            default:
+                return 0;
+        }
+    }
+
+    value_type read_lvt(lvt_reg reg) noexcept override
+    {
+        switch (reg) {
+            case cmci:
+                return msrs::ia32_x2apic_lvt_cmci::get();
+            case timer:
+                return msrs::ia32_x2apic_lvt_timer::get();
+            case thermal:
+                return msrs::ia32_x2apic_lvt_thermal::get();
+            case perf:
+                return msrs::ia32_x2apic_lvt_pmi::get();
+            case lint0:
+                return msrs::ia32_x2apic_lvt_lint0::get();
+            case lint1:
+                return msrs::ia32_x2apic_lvt_lint1::get();
+            case error:
+                return msrs::ia32_x2apic_lvt_error::get();
+
+            default:
+                return 0;
+        }
+    }
+
+    value_type read_count(count_reg reg) noexcept override
+    {
+        switch (reg) {
+            case initial:
+                return msrs::ia32_x2apic_init_count::get();
+            case current:
+                return msrs::ia32_x2apic_cur_count::get();
+
+            default:
+                return 0;
+        }
+    }
+
+    value_type read_div_config() noexcept override
+    { return msrs::ia32_x2apic_div_conf::get(); }
 
     //
     // Register writes
     //
-    void write_eoi() noexcept override;
-    void write_tpr(value_type tpr) noexcept override;
-    void write_svr(value_type svr) noexcept override;
-    void write_icr(value_type icr) noexcept override;
-    void write_lvt(lvt_reg reg, value_type val) noexcept override;
-    void write_init_count(value_type count) noexcept override;
-    void write_div_config(value_type config) noexcept override;
+    void write_eoi() noexcept override
+    { msrs::ia32_x2apic_eoi::set(0xFFFFFFFFULL); }
```

Comment:
the value has to be 0

Reply:
actually, you're note above rings a bell, but I know that it only applies to xapic.  So first double check that a nonzero write is supported in xapic, and then lets change the interface to write_eoi(value_type) and the x2apic implementation will just ignore the argument and write zero